### PR TITLE
feat: add end screen

### PR DESCRIPTION
**Pull request by itsKhadeer**

### CODE REVIEW

**Comment by sparshg** [comment]
~CI is broken~, I will review soon

File: src/main.rs [hunk 1]
```diff
@@ -192,19 +193,42 @@ async fn on_connect(socket: SocketRef, Data(auth): Data<AuthPayload>, pool: Stat
...
                     "attacked",
-                    serde_json::json!({"by": socket.id.as_str(), "at": [i, j], "hit": hit, "sunk": sunk}),
+                    serde_json::json!(
+                        {
+                            "by": socket.id.as_str(),
+                            "at": [i, j],
+                            "hit": hit,
+                            "sunk": sunk,
+                            "winner": winner_loser.0,
+                            "loser": winner_loser.1
```
Comment (sparshg) [inline src/main.rs]: We can just emit `has_won` which can be a bool. Frontend can determine if its them or not using `by` value.

File: src/main.rs [hunk 2]
```diff
@@ -218,6 +242,45 @@ async fn on_connect(socket: SocketRef, Data(auth): Data<AuthPayload>, pool: Stat
         },
     );
 
+    socket.on(
+        "play-again",
```
Comment (sparshg) [inline src/main.rs]: We probably don't need this. If user clicks play again, on frontend, we can create `new State()` and then call `joinRoom()` with the previous room code. The board will automatically reset, and the socket id will be updated automatically on connection.

**Comment by sparshg** [comment]
Thank you so much for contributing!

I think the concept of host is adding too much complexity, since there are only 2 people, both can have equal powers (to start the new game).

To simplify we can just show this same popup to both user. That way we don't have to change the db schema and add host logic. What do you think?
 
image

**Comment by sparshg** [comment]
Thank you so much for contributing!

I think the concept of host is adding too much complexity, since there are only 2 people, both can have equal powers (to start the new game).

To simplify we can just show this same popup to both user. That way we don't have to change the db schema and add host logic. What do you think?
 
image

(Also CI isn't broken, you just need to `cargo sqlx prepare` to update the `.sqlx` folder when the schema is changed)

File: src/main.rs (hunk 2 above)
Comment (itsKhadeer) [reply src/main.rs]: ok

File: src/main.rs (hunk 1 above)
Comment (itsKhadeer) [reply src/main.rs]: cool

**Comment by itsKhadeer** [comment]
> Thank you so much for contributing!
> 
> I think the concept of host is adding too much complexity, since there are only 2 people, both can have equal powers (to start the new game).
> 
> To simplify we can just show this same popup to both user. That way we don't have to change the db schema and add host logic. What do you think?
> 
> image
> (Also CI isn't broken, you just need to `cargo sqlx prepare` to update the `.sqlx` folder when the schema is changed)


I thought in the future we can have more than 2 players in a room like a 4 player game or something.

also the .sqlx is in .gitignore so I didnt' upload the cached queries to github.

if you think the host idea is bad i can drop it. I will change it

**Comment by sparshg** [comment]
> I thought in the future we can have more than 2 players in a room like a 4 player game or something.

Hmm, idk about that, since battleship can't be played by more than 2 players, like chess. Maybe we can introduce that in future if required

> also the .sqlx is in .gitignore so I didnt' upload the cached queries to github.

Its not in the .gitignore file, its used by the CI

> if you think the host idea is bad i can drop it. I will change it

It would be better to do that in its separate PR if required, since there needs to be lot more changes to consider then. Like re-assigning host if the host reconnects, and many other edge cases probably. That shouldn't be clubbed in this PR atleast.

Thanks for your time again

**Comment by itsKhadeer** [comment]
PR updated with requested changes

**Comment by sparshg** [comment]
Fixed the restart game button, thanks for the rest of it

**Pull request merged**